crates/noirc_frontend/src/hir/def_collector/dc_mod.rs

```diff
@@ -54,15 +57,149 @@ pub fn collect_defs(
 
     collector.collect_globals(context, ast.globals, errors);
 
+    collector.collect_traits(ast.traits, crate_id, errors);
+
     collector.collect_structs(ast.types, crate_id, errors);
 
     collector.collect_type_aliases(context, ast.type_aliases, errors);
 
     collector.collect_functions(context, ast.functions, errors);
 
+    collector.collect_trait_impls(context, ast.trait_impls, errors);
+
     collector.collect_impls(context, ast.impls);
 }
 
+fn check_trait_method_implementation_generics(
+    _generics: &[Ident],
+    _noir_function: &NoirFunction,
+    _trait_name: &str,
+) -> Result<(), DefCollectorErrorKind> {
+    // TODO
+    Ok(())
+}
+
+fn check_trait_method_implementation_parameters(
+    parameters: &Vec<(Ident, UnresolvedType)>,
+    noir_function: &NoirFunction,
+    trait_name: &String,
+) -> Result<(), DefCollectorErrorKind> {
+    let expected_num_parameters = parameters.len();
+    let actual_num_parameters = noir_function.def.parameters.len();
+    if actual_num_parameters != expected_num_parameters {
+        let method_name = noir_function.name();
+        let primary_message = format!(
+            "Mismatch - expected {expected_num_parameters} arguments, but got {actual_num_parameters} of trait `{trait_name}` implementation `{method_name}`");
+        return Err(DefCollectorErrorKind::MismatchTraitSignature {
+            primary_message,
+            secondary_message: "".to_string(),
+            span: noir_function.name_ident().span(),
+        });
+    }
+    for (count, (pattern, typ, _abi_vis)) in noir_function.def.parameters.iter().enumerate() {
+        let (_expected_name, expected_type) = &parameters[count];
+        if typ != expected_type {
+            return Err(DefCollectorErrorKind::MismatchTraitSignature {
+                primary_message: format!(
```

Comment:
This should be a different error kind compared to the last error. We can remove the primary and secondary message fields then as well as those can be formatted in when converting the error to a diagnostic later on.

Reply:
I split the mismatch trait signature into 3 errors:
1. return type does not match
2. parameter count does not match
3. specific parameter type does not match